### PR TITLE
feat: [PIE-8741]: Support for looping strategies as runtime input

### PR DESCRIPTION
**Pull request by shalini-agr**
## Describe your changes

Adding support for looping strategies as runtime input

Recording of the behavior: 

https://user-images.githubusercontent.com/97503816/223038380-d48ba36d-86b3-4b9c-ad59-baf37e4142a5.mov

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44700)

### CODE REVIEW

**Comment by harshjain12** [comment]
approving ci side of code

**Review by harshjain12 — APPROVED** [review]
:lgtm:

> Reviewed 16 of 16 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44700#-:-NPpZLuh1tVGAg4e41EO:bnfp4nl)* status: 24 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): sto.

**Review by archit-harness — APPROVED** [review]
> Reviewed 16 of 16 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44700)* status: Some of the checks are too old. Waiting on team(s): sto.

**Review by Eric761 — APPROVED** [review]

**Review by r0610205 — APPROVED** [review]
> Reviewed 16 of 16 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44700)* status: Some of the checks are too old. Waiting on team(s): sto.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSCommandStepFilterJsonCreator.java [hunk 1]
```diff
@@ -61,13 +62,15 @@ public FilterCreationResponse handleNode(FilterCreationContext filterCreationCon
...
 
-  private void validateStrategy(StrategyConfig strategy) {
-    if (strategy != null
-        && ((ParameterField.isNotNull(strategy.getMatrixConfig()) && strategy.getMatrixConfig().getValue() != null)
-            || (strategy.getParallelism() != null
-                && (strategy.getParallelism().getValue() != null
-                    || strategy.getParallelism().getExpressionValue() != null))
-            || strategy.getRepeat() == null || ParameterField.isNull(strategy.getRepeat().getItems()))) {
+  private void validateStrategy(ParameterField<StrategyConfig> strategy) {
+    if (isNotNull(strategy) && strategy.getValue() != null
```
Comment (r0610205) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSCommandStepFilterJsonCreator.java:66]: This condition is very complex, as a suggestion I'd try to break it down to multiple one, getting harder to follow the logic

Comment (shalini-agr) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSCommandStepFilterJsonCreator.java:66]: @r0610205 , This part of code is not owned by us. @vladimir-peric , can you please take care of this? thanks!

**Comment by vladimir-peric** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSCommandStepFilterJsonCreator.java` line 66 at r3](https://reviewable.io/reviews/harness/harness-core/44700#-NPcXfD95rdl_8JYnthP-r3-66:-NQ-04VO6BPpeIoX91HI:bl0mgmd) ([raw file](https://github.com/harness/harness-core/blob/54453e3544343d1b4b916265b2859428253effff/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSCommandStepFilterJsonCreator.java#L66)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

@r0610205 , This part of code is not owned by us. @vladimir-peric , can you please take care of this? thanks!
</blockquote></details>

@shalini-agr sure, I can refactor this as a part of CDS-52879

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44700)* status: 14 file(s) to review. 1. unresolved discussions. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSCommandStepFilterJsonCreator.java` line 66 at r3](https://reviewable.io/reviews/harness/harness-core/44700#-NPcXfD95rdl_8JYnthP-r3-66:-NQ-7XlUCQRIviOdhQcH:bpbgxu4) ([raw file](https://github.com/harness/harness-core/blob/54453e3544343d1b4b916265b2859428253effff/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSCommandStepFilterJsonCreator.java#L66)):*
<details><summary><i>Previously, vladimir-peric (Vladimir Peric) wrote…</i></summary><blockquote>

@shalini-agr sure, I can refactor this as a part of CDS-52879
</blockquote></details>

sure, thanks Vlad.

**Pull request merged**